presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java

```diff
@@ -554,17 +554,38 @@ private PlanBuilder aggregate(PlanBuilder subPlan, QuerySpecification node)
         }
 
         // 4. Project and re-write all grouping functions
+        return handleGroupingOperations(subPlan, node, groupIdSymbol);
+    }
+
+    private PlanBuilder handleGroupingOperations(PlanBuilder subPlan, QuerySpecification node, Optional<Symbol> groupIdSymbol)
+    {
         if (analysis.getGroupingOperations(node).isEmpty()) {
             return subPlan;
         }
-        return subPlan.appendProjections(
-                analysis.getGroupingOperations(node).stream().map(Expression.class::cast).collect(toImmutableList()),
-                symbolAllocator,
-                idAllocator,
-                Optional.of(node),
-                Optional.of(analysis),
-                Optional.of(metadata),
-                groupIdSymbol);
+
+        TranslationMap translations = subPlan.copyTranslations();
+
+        Assignments.Builder projections = Assignments.builder();
+
+        // Add an identity projection for the underlying plan
+        for (Symbol symbol : subPlan.getRoot().getOutputSymbols()) {
+            projections.put(symbol, symbol.toSymbolReference());
+        }
+
+        ImmutableMap.Builder<Symbol, Expression> newTranslations = ImmutableMap.builder();
```

Comment:
Not strictly needed. However, appending new translations at the end means that a particular expression won't be rewritten in terms of a translation that was just added. Consider a query where `grouping(a)` is in the select list and in the having clause. The first projection will map `symbol_a` to `grouping_expression` and on the second projection `symbol_b` will map to `symbol_a`, which is incorrect. To fix this I added an if statement that prevents the projection of multiple equal grouping expressions. Let me know if you think that's incorrect or if it might cause problems later on.